cloudfoundry-client-lib/src/main/java/org/cloudfoundry/client/lib/domain/SecurityGroupRule.java

```diff
@@ -1,88 +1,29 @@
- * You may obtain a copy of the License at
- *
- *      http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
 package org.cloudfoundry.client.lib.domain;
 
-/**
- * A single rule within a security group. See <a href="http://docs.cloudfoundry.org/adminguide/app-sec-groups.html">
- * http://docs.cloudfoundry.org/adminguide/app-sec-groups.html</a> for more details.
- *
- * @author David Ehringer
- * @see http://docs.cloudfoundry.org/adminguide/app-sec-groups.html
- */
-public class SecurityGroupRule {
-
-    private Integer code;
-    private String destination;
-    private Boolean log;
-    private String ports;
-    private String protocol;
-    private Integer type;
-
-    // Required by Jackson.
-    public SecurityGroupRule() {
-    }
-
-    public SecurityGroupRule(String protocol, String ports, String destination) {
-        this(protocol, ports, destination, null);
-    }
+import org.cloudfoundry.client.lib.domain.annotation.Nullable;
+import org.immutables.value.Value;
 
-    public SecurityGroupRule(String protocol, String ports, String destination, Boolean log) {
-        this(protocol, ports, destination, log, null, null);
-    }
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 
-    /**
-     * @param protocol network protocol (tcp,icmp,udp,all)
-     * @param ports port or port range (applicable for tcp,udp,all), may be conditionally <code>null</code>
-     * @param destination destination CIDR or destination range
-     * @param log enables logging for the egress rule, may be <code>null</code>
-     * @param type control signal for icmp, may be <code>null</code>
-     * @param code control signal for icmp, may be <code>null</code>
-     */
-    public SecurityGroupRule(String protocol, String ports, String destination, Boolean log, Integer type, Integer code) {
-        this.protocol = protocol;
-        this.ports = ports;
-        this.destination = destination;
-        this.log = log;
-        this.type = type;
-        this.code = code;
-    }
+@Value.Immutable
+@JsonSerialize(as = ImmutableSecurityGroupRule.class)
+@JsonDeserialize(as = ImmutableSecurityGroupRule.class)
+public interface SecurityGroupRule {
 
-    public Integer getCode() {
-        return code;
-    }
+    String getProtocol();
 
-    public String getDestination() {
-        return destination;
-    }
+    String getPorts();
 
-    public Boolean getLog() {
-        return log;
-    }
+    String getDestination();
 
-    public String getPorts() {
-        return ports;
-    }
+    @Nullable
+    Boolean getLog();
```

Comment:
In theory yes - that's how it should've been. In the current implementation - based on how the class was used before being generated by Immutables. For example, `SecurityGroupRule`'s constructors all required `protocol`, `ports` and `destination` and `null` was never given as a value.
Why? Because if I had checked the CF API and had modeled the POJOs based on it, a good percentage of our tests would've failed, because they don't set a lot of non-nullable fields.